### PR TITLE
Added function GetThreadCpuTimeSeconds in common/time.h

### PR DESCRIPTION
**Pull request by spielawa**
This will be needed for measuring SLAM performance.

### CODE REVIEW

File: cartographer/common/time.h [hunk 1]
```diff
@@ -60,6 +60,8 @@ int64 ToUniversal(Time time);
 // For logging and unit tests, outputs the timestamp integer.
 std::ostream& operator<<(std::ostream& os, Time time);
 
+double GetThreadCpuTimeSeconds();
```
Comment (wohe) [inline cartographer/common/time.h]: Please add a comment what this returns: CPU time consumed by the thread so far.

File: cartographer/common/time.cc [hunk 2]
```diff
@@ -50,5 +51,12 @@ common::Duration FromMilliseconds(const int64 milliseconds) {
       std::chrono::milliseconds(milliseconds));
 }
 
+double GetThreadCpuTimeSeconds() {
+  struct timespec tm;
+  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &tm);
+  return static_cast<double>(tm.tv_sec) +
```
Comment (wohe) [inline cartographer/common/time.cc]: The `static_cast<double>` is not really needed and could be removed.

File: cartographer/common/time.cc [hunk 3]
```diff
@@ -50,5 +51,12 @@ common::Duration FromMilliseconds(const int64 milliseconds) {
       std::chrono::milliseconds(milliseconds));
 }
 
+double GetThreadCpuTimeSeconds() {
+  struct timespec tm;
```
Comment (wohe) [inline cartographer/common/time.cc]: Why `tm`? `thread_cpu_time`?

File: cartographer/common/time.cc [hunk 4]
```diff
@@ -50,5 +51,11 @@ common::Duration FromMilliseconds(const int64 milliseconds) {
       std::chrono::milliseconds(milliseconds));
 }
 
+double GetThreadCpuTimeSeconds() {
+  struct timespec thread_cpu_time;
+  clock_gettime(CLOCK_THREAD_CPUTIME_ID, &thread_cpu_time);
+  return thread_cpu_time.tv_sec + 1e-9 * thread_cpu_time.tv_nsec;
```
[the remark this answers is not in the document]
Comment (wohe) [reply cartographer/common/time.cc]: `1e-9` is `double`, so this is fine. Single precision would be `1e-9f`.

[the remark this answers is not in the document]
Comment (CodeArno) [reply cartographer/common/time.cc]: Ah, you're right.

File: cartographer/common/time.cc [hunk 5]
```diff
@@ -15,7 +15,9 @@
  */
 
 #include "cartographer/common/time.h"
+#include "glog/logging.h"
```
Comment (danielsievers) [inline cartographer/common/time.cc]: +#include <cerrno>
+#include <cstring>

File: cartographer/common/time.cc [hunk 6]
```diff
@@ -50,5 +52,13 @@ common::Duration FromMilliseconds(const int64 milliseconds) {
       std::chrono::milliseconds(milliseconds));
 }
 
+double GetThreadCpuTimeSeconds() {
+  struct timespec thread_cpu_time;
+  if (clock_gettime(CLOCK_THREAD_CPUTIME_ID, &thread_cpu_time) == -1) {
```
Comment (CodeArno) [inline cartographer/common/time.cc]: There are several error codes, e.g. see: https://linux.die.net/man/3/clock_gettime
Checking "if (!clock_gettime() != 0)" is safer.

Comment (spielawa) [reply cartographer/common/time.cc]: "clock_gettime(), clock_settime() and clock_getres() return 0 for success, or -1 for failure (in which case errno is set appropriately)."

File: cartographer/common/time.cc [hunk 7]
```diff
@@ -15,7 +15,11 @@
  */
 
 #include "cartographer/common/time.h"
+#include "glog/logging.h"
```
Comment (wohe) [inline cartographer/common/time.cc]: This belongs after the system includes.

File: cartographer/common/time.cc (hunk 6 above)
Comment (wohe) [reply cartographer/common/time.cc]: +1. Let's check `!= 0`which is more common. Also: Use a `CHECK`? Right now we return bogus values on error.

Comment (spielawa) [reply cartographer/common/time.cc]: ok :) done.

File: cartographer/common/time.cc [hunk 8]
```diff
@@ -50,5 +55,12 @@ common::Duration FromMilliseconds(const int64 milliseconds) {
       std::chrono::milliseconds(milliseconds));
 }
 
+double GetThreadCpuTimeSeconds() {
+  struct timespec thread_cpu_time;
+  CHECK(clock_gettime(CLOCK_THREAD_CPUTIME_ID, &thread_cpu_time) == 0)
+      << "clock_gettime failed: " << std::strerror(errno);
```
Comment (wohe) [inline cartographer/common/time.cc]: `CHECK` should already print the precise call that failed, does it not? So the first part `"clock_gettime failed: "` is redundant. Remove?

Comment (spielawa) [reply cartographer/common/time.cc]: Good point. Done.

**Comment by wohe** [comment]
@wally-the-cartographer merge

**Pull request merged**